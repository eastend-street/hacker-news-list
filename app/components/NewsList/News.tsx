'use client';
import React, { useEffect, useRef } from 'react';
import { Story } from '../../types/story';

export default function News({
  story,
  index,
  isLastStory,
  fetchNextStories,
}: {
  story: Story;
  index: number;
  isLastStory: boolean;
  fetchNextStories: () => void;
}) {
  const newsRef = useRef(null);

  useEffect(() => {
    if (!newsRef?.current || !isLastStory) return;
    const observer = new IntersectionObserver(([entry]) => {
      if (entry.isIntersecting) {
        fetchNextStories();
        observer.unobserve(entry.target);
      }
    });
    observer.observe(newsRef.current);
  }, [isLastStory]);

  return (
    <a
      href={story?.url || `https://news.ycombinator.com/item?id=${story.id}`}
      target="_blank"
      rel="noreferrer"
      ref={newsRef}
      className="hover:underline"
    >
      {index + 1}. {story.title}
    </a>
  );
}
