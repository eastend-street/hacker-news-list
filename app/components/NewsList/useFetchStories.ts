import { useState, useEffect, useCallback } from 'react';
import { HACKER_NEWS_API_BASE_URL } from '../../constants/apiUrls';
import { Story } from '../../types/story';

export default function useFetchStories({
  topStoryIds,
}: {
  topStoryIds: number[];
}) {
  const [stories, setStories] = useState<Story[]>([]);
  const fetchStoryById = useCallback(
    async (storyId: number): Promise<Story> => {
      const response = await fetch(
        `${HACKER_NEWS_API_BASE_URL}/v0/item/${storyId}.json`
      );
      return response.json();
    },
    []
  );

  const fetchStories = useCallback(
    (startIndex: number) => {
      if (startIndex >= topStoryIds.length) return;
      Promise.all(
        topStoryIds
          .slice(startIndex, startIndex + 100)
          .map((storyId) => fetchStoryById(storyId))
      ).then((stories) => {
        setStories((prevState) => [...prevState, ...stories]);
      });
    },
    [topStoryIds, fetchStoryById, setStories]
  );

  useEffect(() => {
    fetchStories(0);
  }, [topStoryIds, fetchStories]);

  return {
    stories,
    fetchStories,
  };
}
