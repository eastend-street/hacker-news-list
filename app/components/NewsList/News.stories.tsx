import React from 'react';
import { Meta, StoryFn } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { Story } from '../../types/story';
import News from './News';

const SampleStory: Story = {
  by: 'fathyb',
  descendants: 118,
  id: 34547259,
  kids: [
    34551288, 34551446, 34551381, 34550957, 34551363, 34559705, 34555437,
    34551359, 34555330, 34553959, 34555234, 34551216, 34553932, 34554496,
    34550988, 34551329, 34556751, 34552308, 34554057, 34551839, 34555521,
    34551266, 34554867, 34555869, 34558878, 34565307, 34554879, 34554925,
    34551236, 34554295, 34560377, 34556853, 34552908, 34554839, 34555318,
    34555085, 34552511, 34554302, 34555641, 34555784, 34555951, 34552526,
    34556686, 34556041, 34551835, 34551513,
  ],
  score: 1137,
  time: 1674834120,
  title: 'Forking Chrome to render in a terminal',
  type: 'story',
  url: 'https://fathy.fr/carbonyl',
};

const SampleStoryWithoutUrl: Story = {
  by: 'apatheticonion',
  descendants: 27,
  id: 34573675,
  kids: [
    34573852, 34573751, 34573900, 34574045, 34574025, 34573767, 34574084,
    34573777, 34574152, 34573846, 34573756, 34573933, 34574156, 34574063,
  ],
  score: 13,
  text: 'In an effort to avoid food poisoning from the more persistent pathogens (like salmonella, e.coli, and other such stomach-acid-surviving bugs), why aren&#x27;t short bursts of radiation used to sterilize raw foods?<p>We go to great lengths to ensure a safe food supply, like processors washing chicken in bleach, washing eggs and massive doses of antibiotics. Despite these efforts, outbreaks are not uncommon. Even vegetables like bean sprouts and onions are commonly associated with outbreaks.<p>Made me wonder, why don&#x27;t we just blast food with a short burst of radiation before&#x2F;after packing?<p>Or if it is used, why isn&#x27;t it more common?',
  time: 1675033047,
  title:
    "Ask HN: Why isn't radiation used to sterilize raw foods (like sashimi, eggs)?",
  type: 'story',
};

export default {
  title: 'components/News',
  component: News,
  args: {
    story: SampleStory,
    index: 0,
    isLastStory: false,
    fetchNextStories: action('fetchNextStories is called'),
  },
} as Meta;

const Template: StoryFn<typeof News> = (args) => <News {...args} />;

export const Default = Template.bind({});

export const WithoutUrl = Template.bind({});
WithoutUrl.args = {
  story: SampleStoryWithoutUrl,
};

export const LastStory = Template.bind({});
LastStory.args = {
  isLastStory: true,
};
