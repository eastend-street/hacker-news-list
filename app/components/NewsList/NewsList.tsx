'use client';
import React from 'react';
import News from './News';
import useFetchStories from './useFetchStories';

export default function NewsList({ topStoryIds }: { topStoryIds: number[] }) {
  const { stories, fetchStories } = useFetchStories({ topStoryIds });

  return (
    <ul>
      {stories.map((story, index) => (
        <li key={story.id} className="my-2">
          <News
            story={story}
            index={index}
            isLastStory={index === stories.length - 1}
            fetchNextStories={() => fetchStories(index + 1)}
          />
        </li>
      ))}
    </ul>
  );
}
