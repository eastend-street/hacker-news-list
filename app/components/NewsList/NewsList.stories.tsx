import React from 'react';
import { Meta, StoryFn } from '@storybook/react';

import NewsList from './NewsList';

const topStoryIds = [34547259, 34546273, 34570264];

export default {
  title: 'components/NewsList',
  component: NewsList,
  args: {
    topStoryIds: topStoryIds,
  },
} as Meta;

const Template: StoryFn<typeof NewsList> = (args) => <NewsList {...args} />;

export const Default = Template.bind({});
