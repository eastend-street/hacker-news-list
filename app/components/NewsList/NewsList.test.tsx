import { render, screen, waitFor, within } from '@testing-library/react';
import { composeStories } from '@storybook/react';

import '@/mocks/intersectionObserverMock';
import * as stories from './NewsList.stories';

const { Default } = composeStories(stories);

describe('components/NewsList', () => {
  test('should render NewsList component', async () => {
    const { container } = render(<Default />);

    await waitFor(() => {
      const list = screen.getByRole('list');
      const { getAllByRole } = within(list);
      const items = getAllByRole('listitem');
      expect(items.length).toBe(3);
    });
    expect(container).toMatchSnapshot();
  });

  test('should render 3 top stories', async () => {
    render(<Default />);
    await waitFor(() => {
      [
        '1. Forking Chrome to render in a terminal',
        '2. Evidence of Multicellularity in Single Cell Organism',
        '3. For longevity, muscle strength may be as important as aerobic exercise',
      ].forEach((title) => {
        const newsAnchorElement = screen.getByRole('link', {
          name: title,
        });
        expect(newsAnchorElement).toBeInTheDocument();
      });
    });
  });
});
