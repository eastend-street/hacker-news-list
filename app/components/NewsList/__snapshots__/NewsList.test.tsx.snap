// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/NewsList should render 3 top stories 1`] = `
<div>
  <ul>
    <li
      class="my-2"
    >
      <a
        class="hover:underline"
        href="https://fathy.fr/carbonyl"
        rel="noreferrer"
        target="_blank"
      >
        1
        . 
        Forking Chrome to render in a terminal
      </a>
    </li>
    <li
      class="my-2"
    >
      <a
        class="hover:underline"
        href="https://research.gatech.edu/more-e-coli-meets-eye-bme-lab-discovers-evidence-multicellularity-single-cell-organism"
        rel="noreferrer"
        target="_blank"
      >
        2
        . 
        Evidence of Multicellularity in Single Cell Organism
      </a>
    </li>
    <li
      class="my-2"
    >
      <a
        class="hover:underline"
        href="https://www.washingtonpost.com/wellness/2023/01/29/strength-training-all-ages/"
        rel="noreferrer"
        target="_blank"
      >
        3
        . 
        For longevity, muscle strength may be as important as aerobic exercise
      </a>
    </li>
  </ul>
</div>
`;

exports[`components/NewsList should render NewsList component 1`] = `
<div>
  <ul>
    <li
      class="my-2"
    >
      <a
        class="hover:underline"
        href="https://fathy.fr/carbonyl"
        rel="noreferrer"
        target="_blank"
      >
        1
        . 
        Forking Chrome to render in a terminal
      </a>
    </li>
    <li
      class="my-2"
    >
      <a
        class="hover:underline"
        href="https://research.gatech.edu/more-e-coli-meets-eye-bme-lab-discovers-evidence-multicellularity-single-cell-organism"
        rel="noreferrer"
        target="_blank"
      >
        2
        . 
        Evidence of Multicellularity in Single Cell Organism
      </a>
    </li>
    <li
      class="my-2"
    >
      <a
        class="hover:underline"
        href="https://www.washingtonpost.com/wellness/2023/01/29/strength-training-all-ages/"
        rel="noreferrer"
        target="_blank"
      >
        3
        . 
        For longevity, muscle strength may be as important as aerobic exercise
      </a>
    </li>
  </ul>
</div>
`;
